src/compiler/checker.ts

```diff
@@ -20233,16 +20238,35 @@ namespace ts {
                         mapper = createTypeMapper(typeParameters, typeArguments);
                     }
                     const typeArgument = typeArguments[i];
+                    const instantiatedConstraint = instantiateType(constraint, mapper);
+                    if (constraints) constraints[i] = instantiatedConstraint;
                     result = result && checkTypeAssignableTo(
                         typeArgument,
-                        instantiateType(constraint, mapper),
+                        instantiatedConstraint,
                         typeArgumentNodes[i],
                         Diagnostics.Type_0_does_not_satisfy_the_constraint_1);
                 }
             }
             return result;
         }
 
+        function getTypeArgumentConstraint(node: TypeNode): Type | undefined {
+            const typeReferenceNode = tryCast(node.parent, isTypeReferenceNodeOrExpressionWithTypeArguments);
+            if (!typeReferenceNode) return undefined;
+            const { typeArguments } = typeReferenceNode;
+            const type = getTypeFromTypeReference(typeReferenceNode);
+            if (type === unknownType) return undefined;
+
+            const typeParameters = getTypeParametersFromTypeReference(typeReferenceNode, type);

```

Comment:
This function reports errors, which is kind of weird; I don't believe we have any other points in the checker API that could report errors as a side effect.

Reply:
Added a `reportErrors` parameter.